### DISCUSSION TITLE
How do I check if the user is logged in without the SDK?

### DISCUSSION THREAD

**Discussion by velllu** [pocketbase/pocketbase#2998] [Q&A]
I added a middleware in my server that checks if you have a cookie with a token, how do I check if the token is valid? There's no auth_with_token function

**Comment by ganigeorgiev** [comment]
> I added a middleware in my server ...

Do you mean custom echo middleware or some external application (eg. node) middleware?

If you are using PocketBase as framework, to check whether a token is valid, you can call:
```js
record, err := app.Dao().FindAuthRecordByToken(
    token,
    app.Settings().RecordAuthToken.Secret,
)
```

**Reply by velllu** [reply]
I am not using node/js at all, by middleware, I just mean something that gets run before any request, to check if the user is logged in.

Do you mind explaining what that snippet does and what "Dao" is? 🤔

**Reply by ganigeorgiev** [reply]
The above code sample will work only if you [use PocketBase as framework](https://pocketbase.io/docs/use-as-framework/).

> by middleware, I just mean something that gets run before any request, to check if the user is logged in.

But how and where? If you just want to check whether a token is valid you can send `Authorization: TOKEN` header with `POST /api/collections/users/auth-refresh` request (see https://pocketbase.io/docs/api-records/#auth-record-actions). On success it will return the latest user data with renewed token, otherwise - error response.

**Reply by velllu** [reply]
I am using axum-rs, a middleware is just a function where you can interact with the request before actually forwarding the request to the client.

What I am looking to do is add protected routes, this is how I planned to do it:

- Check the cookies of the request
- If there is a cookie with the token, and the token is the right one, then everything proceeds as planned
- If the token is not valid/outdated/not existing, it will send a redirect to the /register page.


This can't really be done if the token is refreshed :thinking:

**Reply by ganigeorgiev** [reply]
PocketBase APIs are fully stateless and we don't have sessions in the traditional sense (the tokens are also not stored anywhere). A request is considered authenticated if it has a valid `Authorization: TOKEN` header with it.

Tokens are invalidated when:
- they expire (aka. the `exp` claim in the JWT payload)
- the user specific token secret (_stored as `tokenKey` in the auth colleciton table_) is changed (eg. on password reset)
- the app shared token secret is changed (this invalidates all tokens from a single type for all users)

> This can't really be done if the token is refreshed...

When you call `POST /api/collections/users/auth-refresh` it **doesn't invalidate the old token**. It returns a new one with updated `exp` claim and up-to-date user data. If you don't want to refresh the token that is in your cookie you can safely ignore the returned new one from the response.

**Reply by velllu** [reply]
Is there no way to verify if a user is logged in then? I cannot check if a user has a valid token without making a new one right? How does the import/export from cookie function on the js sdk works then?

**Reply by ganigeorgiev** [reply]
I'm failing to understand what you are trying to do.

> I cannot check if a user has a valid token without making a new one right?

Yes, but why does it matter if the response return token with refreshed `exp` claim? This seems to be a common request but I don't really think that the HS256 token generation has that much impact on the performance, if that's your concern (I'll add a test to the [benchamrks repo](https://github.com/pocketbase/benchmarks) for this soon).

> How does the import/export from cookie function on the js sdk works then?

They create and parse a plain cookie string that contains the token.

Again, PocketBase doesn't have sessions. Each PocketBase request is a standalone and validated separately based on the request `Authorization` header.
